File: algorithms/2_medium/P/00567_permutation_in_string/solution.c
#define INC 1
#define DEC -1

typedef ptrdiff_t isize;

static inline void update_freqs_and_mask(isize* freqs, const isize idx, const isize amount, isize* mask_non_zeros)
{
    freqs[idx] += amount;
    if (freqs[idx] == 0) {
        *mask_non_zeros &= ~(1 << idx);
    } else {
        *mask_non_zeros |= (1 << idx);
    }
}

bool checkInclusion(char* s1, char* s2)
{
    assert(s1 != NULL && s2 != NULL);
    const isize s1_len = (isize)strlen(s1);
    const isize s2_len = (isize)strlen(s2);
    assert(s1_len > 0 && s2_len > 0);
    if (s2_len < s1_len) { return false; }

    isize freqs[26] = { 0 };
    isize mask_non_zeros = 0;

    for (isize i = 0; i < s1_len; i += 1) {
        assert(s1[i] >= 'a' && s1[i] <= 'z');
        update_freqs_and_mask(freqs, s1[i] - 'a', DEC, &mask_non_zeros);
    }

    for (isize i = 0; i < s2_len; i += 1) {
        assert(s2[i] >= 'a' && s2[i] <= 'z');
        update_freqs_and_mask(freqs, s2[i] - 'a', INC, &mask_non_zeros);
        if (i >= s1_len) {
            update_freqs_and_mask(freqs, s2[i - s1_len] - 'a', DEC, &mask_non_zeros);
        }

        if (mask_non_zeros == 0) {
            return true;
        }
    }

    return false;
}

File: algorithms/2_medium/P/00567_permutation_in_string/solution.rs
enum IncDec {
    Inc,
    Dec,
}

impl Solution {
    #[inline(always)]
    fn update_freqs_and_mask(freqs: &mut[isize], idx: usize, state: IncDec, mask_non_zeros: &mut isize) {
        freqs[idx] += match state {
            IncDec::Inc => 1,
            IncDec::Dec => -1,
        };
        if freqs[idx] == 0 {
            *mask_non_zeros &= !(1 << idx);
        } else {
            *mask_non_zeros |= (1 << idx);
        }
    }
    pub fn check_inclusion(s1: String, s2: String) -> bool {
        let (b1, b2) = (s1.as_bytes(), s2.as_bytes());
        let (b1_len, b2_len) = (b1.len(), b2.len());
        assert!(b1_len > 0 && b2_len > 0);
        if b2_len < b1_len { return false; }

        let mut freqs = [0_isize; 26];
        let mut mask_non_zeros = 0_isize;

        for &byte in s1.as_bytes() {
            assert!(byte >= b'a' && byte <= b'z');
            Self::update_freqs_and_mask(&mut freqs, (byte - b'a') as usize, IncDec::Dec, &mut mask_non_zeros);
        }

        for i in 0..b2_len {
            let byte_to_insert = b2[i];
            assert!(byte_to_insert >= b'a' && byte_to_insert <= b'z');
            Self::update_freqs_and_mask(&mut freqs, (byte_to_insert - b'a') as usize, IncDec::Inc, &mut mask_non_zeros);
            if i >= b1_len {
                let byte_to_remove = b2[i - b1_len];
                Self::update_freqs_and_mask(&mut freqs, (byte_to_remove - b'a') as usize, IncDec::Dec, &mut mask_non_zeros);
            }

            if mask_non_zeros == 0 {
                return true;
            }
        }

        false
    }
}
